File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/SystemUINetworkSpeedSIzeHook.java
package com.qimian233.ztool.hook.modules.systemui;

import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.RelativeSizeSpan;
import android.widget.TextView;
import com.qimian233.ztool.hook.base.BaseHookModule;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

/**
 * 系统UI网速显示样式Hook模块
 * 修改系统状态栏中的网速显示，使数字部分更大、单位部分更小
 */
public class SystemUINetworkSpeedSIzeHook extends BaseHookModule {

    @Override
    public String getModuleName() {
        return "systemui_network_speed_size";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{"com.android.systemui"};
    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
        String packageName = lpparam.packageName;

        if ("com.android.systemui".equals(packageName)) {
            hookSystemUI(lpparam);
        }
    }

    private void hookSystemUI(XC_LoadPackage.LoadPackageParam lpparam) {
        try {
            log("开始Hook系统UI网速显示");

            // 使用beforeHookedMethod避免递归调用
            XposedHelpers.findAndHookMethod(TextView.class, "setText",
                    CharSequence.class, new XC_MethodHook() {
                        @Override
                        protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                            try {
                                CharSequence text = (CharSequence) param.args[0];

                                // 检查是否是网络速度显示文本
                                if (isNetworkSpeedText(text)) {
                                    CharSequence styledText = createStyledSpeedText(text.toString());
                                    param.args[0] = styledText; // 直接修改参数
                                    log("成功修改网速显示样式");
                                }
                            } catch (Throwable t) {
                                // 忽略处理过程中的异常
                            }
                        }
                    });

            log("系统UI网速显示Hook成功");
        } catch (Throwable e) {
            logError("系统UI网速显示Hook失败", e);
        }
    }

    /**
     * 检查是否是网速文本
     * 网速文本特征：包含换行符，以下载/上传速度单位结尾
     */
    private boolean isNetworkSpeedText(CharSequence text) {
        if (text == null) return false;
        String str = text.toString();
        return str.contains("\n") &&
                (str.endsWith("K/s") || str.endsWith("M/s") || str.endsWith("G/s") ||
                        str.endsWith("KB/s") || str.endsWith("MB/s") || str.endsWith("GB/s") ||
                        str.endsWith("Kbps") || str.endsWith("Mbps") || str.endsWith("Gbps"));
    }

    /**
     * 创建带样式的网速文本
     * 数字部分1.3倍大小，单位部分0.9倍大小
     */
    private CharSequence createStyledSpeedText(String originalText) {
        if (!originalText.contains("\n")) {
            return originalText;
        }

        String[] parts = originalText.split("\n");
        if (parts.length != 2) {
            return originalText;
        }

        String numberPart = parts[0];
        String unitPart = parts[1];

        SpannableString spannableString = new SpannableString(numberPart + "\n" + unitPart);

        // 设置数字部分相对大小为1.3倍（更大）
        spannableString.setSpan(new RelativeSizeSpan(1.3f),
                0, numberPart.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

        // 设置单位部分相对大小为0.9倍（更小）
        spannableString.setSpan(new RelativeSizeSpan(0.9f),
                numberPart.length() + 1, spannableString.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

        return spannableString;
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/statusbar/SystemUINetworkSpeedSizeHook.kt
package com.qimian233.ztool.hook.modules.systemui.statusbar

import android.annotation.SuppressLint
import android.text.SpannableString
import android.text.Spanned
import android.text.style.RelativeSizeSpan
import android.widget.TextView
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam

/**
 * SystemUI network speed display style hook module.
 * Modifies the network speed display in the system status bar so the number part is
 * larger and the unit part is smaller.
 *
 * Detection: hooks [TextView.setText] and uses [Class.isInstance] to check whether the
 * caller is a com.android.systemui.zui.NetworkSpeedView instance. Confirmed via Jadx
 * decompilation, all setText(...) call sites inside that class (direct calls from
 * updateNetworkSpeedViewStatus, internal Handler what==1/what==10 branches) are used
 * only for network speed text, so identifying by call source is more precise and
 * reliable than the previous approach of matching "K/s"/"M/s" string suffixes, and is
 * unaffected by system text format changes.
 */
@SuppressLint("PrivateApi")
class SystemUINetworkSpeedSizeHook : AppHookModule() {

    companion object {
        private const val NETWORK_SPEED_VIEW_CLASS = "com.android.systemui.zui.NetworkSpeedView"
    }

    override fun getModuleName(): String = PreferenceKeys.SYSTEMUI_NETWORK_SPEED_SIZE.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.SYSTEM_UI.packageName)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        try {
            logger.info("Hooking SystemUI network speed display")

            // Load the NetworkSpeedView class for caller type checks in the callback
            val networkSpeedViewClass =
                param.defaultClassLoader.loadClass(NETWORK_SPEED_VIEW_CLASS)

            // Use the hook callback to avoid recursive calls
            val setTextMethod =
                TextView::class.java.getDeclaredMethod("setText", CharSequence::class.java)
            hookWithId(setTextMethod, "set_text") { chain ->
                try {
                    // Only handle NetworkSpeedView instance text, not other status bar TextViews
                    if (networkSpeedViewClass.isInstance(chain.thisObject)) {
                        val text = chain.args[0] as CharSequence
                        if (isNetworkSpeedText(text)) {
                            val styledText = createStyledSpeedText(text.toString())
                            logger.debug("Successfully modified network speed display style")
                            return@hookWithId chain.proceed(arrayOf<Any>(styledText))
                        }
                    }
                } catch (_: Throwable) {
                    // Ignore exceptions during processing
                }
                chain.proceed()
            }

            logger.info("SystemUI network speed display hooks applied")
        } catch (e: Throwable) {
            logger.error("Failed to hook SystemUI network speed display", e)
        }
    }

    /**
     * Check whether this is network speed text.
     * NetworkSpeedView's setText content is always the two-line "number\nunit" format
     * (e.g. "12.3\nK/s"). The call source is already restricted to NetworkSpeedView,
     * so only the newline format requirement is checked here; no dependency on
     * specific unit suffix matching.
     */
    private fun isNetworkSpeedText(text: CharSequence?): Boolean {
        return text != null && text.contains("\n")
    }

    /**
     * Create styled network speed text.
     * Number part at 1.3x size, unit part at 0.9x size.
     */
    private fun createStyledSpeedText(originalText: String): CharSequence {
        if (!originalText.contains("\n")) {
            return originalText
        }

        val parts = originalText.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        if (parts.size != 2) {
            return originalText
        }

        val numberPart = parts[0]
        val unitPart = parts[1]

        val spannableString = SpannableString(numberPart + "\n" + unitPart)

        // Set the number part relative size to 1.3x (larger)
        spannableString.setSpan(
            RelativeSizeSpan(1.3f),
            0, numberPart.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )

        // Set the unit part relative size to 0.9x (smaller)
        spannableString.setSpan(
            RelativeSizeSpan(0.9f),
            numberPart.length + 1, spannableString.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )

        return spannableString
    }
}
